Assets/Labrys/FeatureEditor/FeatureAsset.cs

```diff
@@ -0,0 +1,372 @@
+
+		[SerializeField, HideInInspector]
+		private List<Vector2Int> selectionPositions;
+
+		public void OnAfterDeserialize()
+		{
+			if (sectionPositions.Count != sectionData.Count
+				|| linkPositions.Count != linkData.Count)
+				throw new ArgumentException("Malformed feature data");
+
+			for(int i = 0; i < sectionPositions.Count; i++)
+			{
+				sections.Add(sectionPositions[i], sectionData[i]);
+			}
+
+			for (int i = 0; i < linkPositions.Count; i++)
+			{
+				links.Add(linkPositions[i], linkData[i]);
+			}
+
+			for(int i = 0; i < selectionPositions.Count; i++)
+			{
+				selected.Add(selectionPositions[i]);
+			}
+		}
+
+		public void OnBeforeSerialize()
+		{
+			sectionPositions.Clear();
+			sectionData.Clear();
+			foreach(KeyValuePair<Vector2Int, Section> kvp in sections)
+			{
+				sectionPositions.Add(kvp.Key);
+				sectionData.Add(kvp.Value);
+			}
+
+			linkPositions.Clear();
+			linkData.Clear();
+			foreach (KeyValuePair<Vector2Int, Link> kvp in links)
+			{
+				linkPositions.Add(kvp.Key);
+				linkData.Add(kvp.Value);
+			}
+
+			selectionPositions.Clear();
+			foreach(Vector2Int position in selected)
+			{
+				selectionPositions.Add(position);
+			}
+		}
+		#endregion
+
+		public FeatureAsset()
+		{
+			sections = new Dictionary<Vector2Int, Section>();
+			links = new Dictionary<Vector2Int, Link>();
+			selected = new HashSet<Vector2Int>();
+
+			sectionPositions = new List<Vector2Int>();
+			sectionData = new List<Section>();
+			linkPositions = new List<Vector2Int>();
+			linkData = new List<Link>();
+			selectionPositions = new List<Vector2Int>();
+		}
+
+		public void AddSection(Vector2Int gridPosition)
+		{
+			if(Section.IsValidPosition(gridPosition)
+				&& !sections.ContainsKey(gridPosition))
+			{
+				sections.Add(gridPosition, new Section() { Variant = "default" });
+				UpdateLinks(gridPosition);
+			}
+		}
+
+		public bool RemoveSection(Vector2Int gridPosition)
+		{
+			if(sections.Remove(gridPosition))
+			{
+				UpdateLinks(gridPosition);
+				DeselectSection(gridPosition);
+				return true;
+			}
+			return false;
+		}
+
+		public bool SelectSection(Vector2Int gridPosition)
+		{
+			if(Section.IsValidPosition(gridPosition))
+			{
+				return selected.Add(gridPosition);
+			}
+			return false;
+		}
+
+		public bool IsSelected(Vector2Int gridPosition)
+		{
+			if (Section.IsValidPosition(gridPosition))
+			{
+				return selected.Contains(gridPosition);
+			}
+			return false;
+		}
+
+		public bool DeselectSection(Vector2Int gridPosition)
+		{
+			return selected.Remove(gridPosition);
+		}
+
+		public bool HasSectionAt(Vector2Int gridPosition)
+		{
+			return Section.IsValidPosition(gridPosition) 
+				&& sections.ContainsKey(gridPosition);
+		}
+
+		public bool HasLinkAt(Vector2Int gridPosition)
+		{
+			return !Section.IsValidPosition(gridPosition) 
+				&& links.ContainsKey(gridPosition);
+		}
+
+		public bool TryGetSection(Vector2Int gridPosition, out Section section)
+		{
+			section = default;
```

Comment:
default in this case == null, might be clearer to just write out `null`. First time I read this I thought there was a default non-null initial value for `section`.

Reply:
Oop. Leftover from when Section was briefly a struct. Will change.